code/contracts/ONEWallet.sol

```diff
@@ -7,14 +7,15 @@ contract ONEWallet {
     uint8 interval; // otp interval in seconds, default is 30
     uint32 t0; // starting time block (effectiveTime (in ms) / interval)
     uint32 lifespan;  // in number of block (e.g. 1 block per [interval] seconds)
-    uint8 maxOperationsPerInterval; // number of transactions permitted per OTP interval. Each transaction shall have a unique nonce. The nonce is auto-incremented within each interval
+    uint8 maxOperationsPerInterval; // number of transactions permitted per OTP interval. Each transaction shall have a unique nonce. The nonce is auto-incremented within each interval 
 
     // global mutable
     address payable lastResortAddress; // where money will be sent during a recovery process (or when the wallet is beyond its lifespan)
     uint256 dailyLimit; // uint128 is sufficient, but uint256 is more efficient since EVM works with 32-byte words.
     uint256 spentToday; // note: instead of tracking the money spent for the last 24h, we are simply tracking money spent per 24h block based on UTC time. It is good enough for now, but we may want to change this later.
-    uint32 lastTransferDay;
+    uint32 lastTransferDay; // I'd use uint instead
 
+    // IH: you can drop nonce stuff if you implement commits as mapping (hash => Commit). After reveal you might delete the Commit entry; the timestamp's fresheness will ensure protection against replay in the future.
```

Comment:
It seems still insufficient. Consider the scenario which a commit is revealed during the same 30-second interval, but may span across multiple blocks (recall Harmony generates a block per 2 seconds). If a commit is deleted after the corresponding reveal is completed and there is no nonce tracking, an attacker may immediately recommit another transaction using a correct proof but with modified parameters (e.g. different transfer destination and amount), then immediately reveal it. Since the commit is already deleted and nonce is not used, the smart contract would have no way to identify that this new transaction should be rejected.

Reply:
Yes, you are right. I've somehow already assumed HOTPs, in which this cannot not happen by default since SC remembers ids of the used ones (or id of the last used one) while allowing only not yet used (or later ones). But going back to TOTPs - I'd still suggest to drop the nonce stuff (as too verbose) and do the things simpler. So, to resolve the issue with deleting from the map of commits, I'd use a dedicated cleanup external method (called once in a while), while using another EnumerableMap (EOTPs => timestamps) to track EOTPs to delete, and their timestamps of expiration. In this method, SC would just iterate over EOTPs in the map and check the time validity of EOTP according to block.timestamp, followed by the delete from both maps with the purpose of saving the storage space. Also, in any revealX(), SC would need in addition to insert data into the second map.

Reply:
This is feasible. One issue is the caller of the external method would have to pay for the gas to cleanup. The attacker should pay for that, if an attack exists. But this can be mitigated by making this method internal and call it sporadically as part of the commit process.

But the primary purpose of nonce is to enable multiple operations per interval, not tracking or managing commits. Each nonce value corresponds to an EOTP inside an interval. Each interval has one (T)OTP, but has maxOperationsPerInterval number of EOTPs (corresponding to each nonce). The current method does not track used EOTPs. It simply requires a transaction to use a higher nonce than before (thus must correspond with different EOTP).

The method for managing and tracking commits are independent from nonce.